csrc/multidevice/communicator.h

```diff
@@ -7,84 +14,107 @@
-      RankType server_local_rank = COMM_SERVER_LOCAL_RANK_DEFAULT);
+      CommunicatorBackend backend = comm_backend_default,
+      RankType server_local_rank = comm_server_local_rank_default);
+
+  Communicator(const Communicator&) = delete;
+  Communicator& operator=(const Communicator&) = delete;
 
   // returns if distributed config is available
   auto is_available() const {
     return is_available_;
   }
 
-  // returns the rank of the current process
-  auto rank() const {
-    return rank_;
-  }
-
   // returns the number of processes in the communicator
   auto size() const {
     return size_;
   }
 
-  // returns the local rank of the current process (within the node)
-  auto local_rank() const {
-    return local_rank_;
-  }
-
   // returns the local number of processes in the communicator (within the node)
   auto local_size() const {
     return local_size_;
   }
 
-  // returns the flattenend list of ranks of the communicator
-  auto ranks() const {
-    std::vector<RankType> ret(size());
-    std::iota(ret.begin(), ret.end(), 0);
-    return ret;
-  }
-
-  // performs a (blocking) send/receive p2p data transfer
-  void sendRecv(
-      RankType receiver_rank,
-      RankType sender_rank,
+  // performs a send/receive p2p data transfer
+  c10::intrusive_ptr<c10d::Work> sendRecv(
+      DeviceIdxType receiver,
+      DeviceIdxType sender,
       std::vector<at::Tensor>& tensor,
       int tag = 0);
 
   // performs a blocking barrier in the communicator
   void barrier() const {
-    pg_->barrier()->wait();
+    world_->barrier()->wait();
+  }
+
+  // returns the backend associated with a subset of devices/ranks given as
+  // arguments
+  c10::intrusive_ptr<c10d::Backend> getBackendForTeam(
+      const std::vector<DeviceIdxType>& devices);
+
+  // returns the device associated with the current process
+  auto device() const {
+    return at::Device("cuda:" + std::to_string(local_rank_));
+  }
+
+  // returns the device Id associated with the current process
+  DeviceIdxType deviceId() const {
+    return rankToDiD(rank_);
   }
 
-  // stores the process group backend
  private:
+  // returns the rank corresponding to a device index
+  RankType dIdToRank(DeviceIdxType d_id) const {
+    return static_cast<RankType>(d_id);
+  }
+
+  // returns the device index corresponding to a rank
+  DeviceIdxType rankToDiD(RankType rank) const {
+    return static_cast<DeviceIdxType>(rank);
+  }
+
   bool is_available_;
+  CommunicatorBackend backend_;
   RankType rank_;
   int64_t size_;
   RankType local_rank_;
   int64_t local_size_;
   std::string master_addr_;
   int master_port_;
-  c10::intrusive_ptr<c10d::Backend> pg_;
+  // stores the world's store used for the backend init
+  c10::intrusive_ptr<c10d::TCPStore> store_;
+  // stores the world's backend
+  c10::intrusive_ptr<c10d::Backend> world_;
+  // cache for the created teams
+  std::map<std::vector<RankType>, c10::intrusive_ptr<c10d::Backend>> teams_;
```

Comment:
Should this be `std::vector<DeviceIdxType>`?

Reply:
correct!